### PR TITLE
Fix #6 by adding support for weak references.

### LINKED ISSUE #6: Garbage collection fails for cross-database weak references and legacy weak references
Not all forms of weak references are handled by `getrefs`. Specifically, cross-database weak references use a different format, reversing the arguments of name and oid. This leads to incorrect data in `Bad`, for example:
```
[('Sessions', '\x00\x00\x00\x00\x00\x01\xecb'),
 ('\x00\x00\x00\x00\x00\x042\xd2', 'Users'),
 ('Sessions', '\x00\x00\x00\x00\x00\x01\xecc'),
 ('Sessions', '\x00\x00\x00\x00\x00\x01\xecd')]
```
Note that entry 1 is backwards. This in turn leads to a `KeyError`:
```
  File "//zc.zodbdgc-1.0.0-py2.7.egg/zc/zodbdgc/__init__.py", line 152, in gc
    bad = gc_(close, conf, days, ignore, conf2, fs, untransform, ptid)
  File "//zc.zodbdgc-1.0.0-py2.7.egg/zc/zodbdgc/__init__.py", line 286, in gc_
    if good.insert(*ref) and bad.has(*ref):
  File "//zc.zodbdgc-1.0.0-py2.7.egg/zc/zodbdgc/__init__.py", line 374, in insert
    data = self[name].get(prefix)
KeyError: '\x00\x00\x00\x00\x00\x042\xd2'
```

This is biting one of our production databases. I already have code to deal with this (and the legacy weak reference case) but I forgot to submit it for 1.0.0. I need to write a test case and then open a PR.

### PR DESCRIPTION
**Pull request by jamadden**
Weak references, whether in a single DB or cross-DB, are not considered strong and do not keep an object from being GC'd (just like ZODB.serialize.referencesf).

There are new doctests for this that go into detail about how weak refs and object lifetimes interact.

The changes to `getrefs` are code we'd been using in production for awhile, having encountered most of those cases. The tests exercise most of them too. 

The only change from production is that I'd been treating weakrefs like strong refs and returning them, thus artificially keeping objects alive. Whoops. Hence the comments on how to do that.

It's sort of broken that if you have a weakref and a strongref, both cross-db, and the strongref gets loaded into the PickleCache at the same time as you access the weakref, you get a POSKeyError instead of the expected return value of None. But that's not a problem that can be solved here, I think.

### CODE REVIEW

**Comment by jamadden** [comment]
Is anyone interested in reviewing? @mgedmin or @jimfulton perhaps? I'm hoping to merge and release this soon-ish.

**Comment by jimfulton** [comment]
I'll look this weekend.

**Comment by mgedmin** [comment]
Since this PR is the first time I heard about ZODB being able to store weak refs, I don't feel qualified to review.

**Comment by jamadden** [comment]
The change itself is pretty simple and based on the documentation in [ZODB.serialize](https://github.com/zopefoundation/ZODB/blob/master/src/ZODB/serialize.py#L80). For the previously handled reference types, nothing changes. The 'w' reference type is now ignored; previously it would either yield a ValueError (because it returned too few values to unpack) or a KeyError (IIRC) because it returned a database name string instead of an OID. There's two new branches, one for a corrupted reference type that we once saw, and one to future proof against new reference types so that silent corruption doesn't happen.

**Comment by jimfulton** [comment]
I'll review this this weekend.

File: src/zc/zodbdgc/README.test [hunk 1]
```diff
@@ -726,6 +726,219 @@ files directly:
...
+    >>> zc.zodbdgc.gc('config', days=2, return_bad=True)
+    db1: roots
+    db1: recent
+    db2: roots
+    db2: recent
+    db1: remove garbage
+    Removed 1 objects from db1
+    db2: remove garbage
+    Removed 0 objects from db2
+    [('db1', 1)]
```
Comment (jimfulton) [inline src/zc/zodbdgc/README.test]: Wait, why doesn't the reference from the root object in database 2 keep the object in database 1 around?

Comment (jamadden) [inline src/zc/zodbdgc/README.test]: Sadly this looks like a bug in the gc algorithm. The item in question, OID 1 in db1, is found in the Bad object after all references have been collected, so it is removed. I added some logging to both `oidset.insert` and `Bad.insert` to see why this would be. Here's the logging for this example:
```
    db1: roots
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01',)] db1 ()
    db1: recent
    gc_: Has data <ZODB.FileStorage.FileStorage.Record object at 0x10384bc10> 'cpersistent.mapping\nPersistentMapping\nq\x01.}q\x02U\x04dataq\x03}q\x04s.'
    BAD: Inserting db1 '\x00\x00\x00\x00\x00\x00\x00\x01' '\x03}\xfe\xe7\\8F\x99' <generator object getrefs at 0x103852780>
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01',)] db1 ()
    db2: roots
    ['m', ('db1', '\x00\x00\x00\x00\x00\x00\x00\x01', None)] db2 ()
    GOOD Insert db1 '\x00\x00\x00\x00\x00\x00\x00\x01'
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01', 'db1')] db2 ()
    db2: recent
    ['m', ('db1', '\x00\x00\x00\x00\x00\x00\x00\x01', None)] db2 ()
    GOOD Insert db1 '\x00\x00\x00\x00\x00\x00\x00\x01'
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01', 'db1')] db2 ()
    db1: remove garbage
    Remove '\x00\x00\x00\x00\x00\x00\x00\x01' '\x03}\xfe\xe7\\8F\x99' db1
    Removed 1 objects from db1
    db2: remove garbage
    Removed 0 objects from db2
    [('db1', 1)]
```

You can see that OID 1 is inserted into Bad, although it is later inserted into the good oidset. Because it's found in Bad, it is removed.

Here's the relevant code from the `gc_` function that prints "Has data" and inserts the item into Bad:

```python
        for trans in it:
            for record in trans:
                ...

                oid = record.oid
                data = record.data
                if data:
                    if deleted.has(name, oid):
                        continue
                    if good.has(name, oid):
                       ... elide 2...
                    else:
                        if DEBUG and oid == '\x00\x00\x00\x00\x00\x00\x00\x01':
                            print("Has data", record, repr(data))
                        bad.insert(name, oid, record.tid,
                                   getrefs(data, name, ignore))

                else:
                    # deleted record
                    ...
```

In other words, we encounter the record in DB1 for OID 1. It's not deleted, and it's not known to be good (have other references), so it must be Bad. Even though traversing through DB2 later reveals it to be good, it never gets removed from Bad. 

I *think* this is because there are no references in its "home" database DB1, so it never gets put into the good set in the first place, and so the code that wants to pop from bad (... elide 2 ..., about line 278) never gets a chance to do so. But I'm not entirely certain.

Comment (jimfulton) [inline src/zc/zodbdgc/README.test]: Yes, you've encountered a huge bug. :(

The code that sets the roots and their references and that add new records assumes that there aren't bad yet, but that's not consistent with the way the loops are organized.

I think we need to fix this on another branch, which will break your test. :)

The fix is straightforward.  Just iterate over all databases to insert the roots, ther refs and new objects, and then iterate over the databases to handle old records.

This test provides a good start on a test for the fix. Do you feel like tackling it? Or shall I?

Comment (jamadden) [inline src/zc/zodbdgc/README.test]: I'm happy to take a pass at it. I'll probably start from here.

Comment (jamadden) [inline src/zc/zodbdgc/README.test]: > The fix is straightforward. Just iterate over all databases to insert the roots, ther refs and new objects, and then iterate over the databases to handle old records.

Can it be simpler? Just check if the name,oid is in `good` before removing it? Like so:
```
        for oid, tid in bad.iterator(name):
            if good.has(name, oid): # ADD THIS
                continue
            try:
                storage.deleteObject(oid, tid, t)
            except (ZODB.POSException.POSKeyError,
                    ZODB.POSException.ConflictError):
                continue
```

If I add a test case for this scenario, that change makes that test pass (and breaks my original test :)

Comment (jimfulton) [inline src/zc/zodbdgc/README.test]: While this fix is simpler, it has the disadvantage of storing the oid twice.  A challenge for this algorithm is that it's pretty memory intensive for large databases and I'd like to mitigate that.

My fix is pretty simple too. You just have to break the oouter database loop in 2. Do the roots and new records for all of the databases, and then do the old records for all the databases.

**Comment by jimfulton** [comment]
As an aside, I have ideas for a next-generation file-storage, https://github.com/jimfulton/filestorage2 that can address a lot of scalability issues wrt packing and gc.

File: src/zc/zodbdgc/README.test (hunk 1 above)
Comment (jamadden) [inline src/zc/zodbdgc/README.test]: The test is at https://github.com/zopefoundation/zc.zodbdgc/compare/incorrect-remove#diff-e04096299edfbbb8f84b42d53d6eace3R487

Comment (jamadden) [inline src/zc/zodbdgc/README.test]: I had thought there might be a problem with breaking the loops up if days=0, but that doesn't seem to be the case (it still works). I'm updating the PR.

**Comment by jamadden** [comment]
This PR has been superseded by #8.
